Cmake + workflows

wkpark opened the pull request:
CmakeList.txt + github workflows action

* CMake by @Jamezo97
* github workflows actions - support ubuntu and windows (windows disabled for now)
* cu11.8, cu12.1 + py310, py311 build matrix

Please see ~~https://github.com/wkpark/bitsandbytes/actions/runs/7166014267~~
~~https://github.com/wkpark/bitsandbytes/actions/runs/7167362163~~
https://github.com/wkpark/bitsandbytes/actions/runs/7648577944

TimDettmers commented:
Thank you! We are planning to switch to make, and this is a great PR!

I do not have an overview of how this is different or could benefit from #220 and #229. Could you please have a look and discuss the differences and if we could benefit from some of the old code. Thank you!

rickardp commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,161 @@
+
+      matrix:
+        os: [ubuntu-latest]
+        python-version: ['3.10', '3.11']
```
I would recommend building only the native library here using a build matrix, then a separate job to build the Python binary wheels.

rickardp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+set_target_properties(libbitsandbytes
+    PROPERTIES
+        OUTPUT_NAME ${BNB_OUTPUT_NAME}
+        # We have to use a generator expression to prevent MSVC Debug/Release subdirs being made
```
As an alternative, you can do something like

```
if(MSVC)
set_target_properties(bitsandbytes PROPERTIES LIBRARY_OUTPUT_DIRECTORY_RELEASE bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES LIBRARY_OUTPUT_DIRECTORY_DEBUG bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES RUNTIME_OUTPUT_DIRECTORY_RELEASE bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES RUNTIME_OUTPUT_DIRECTORY_DEBUG bitsandbytes)
endif()

```

Possibly more cmake-canonical, not sure which is the best

wkpark replied:
thank you for your information. I've missed `LIBRARY_OUTPUT_DIRECTORY` property!

rickardp commented:
Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling.

wkpark commented:
> Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling. Unfortunately the original PRs (#220 etc) cannot be viewed because of force pushes. I'll see if I can dig up a relevant diff

Windows specific stuff extracted on PR #873 and #876
See also https://github.com/wkpark/bitsandbytes/actions (several times compiled, partial squashed rebased blah blah..)

(and pytest stuff is not fixed/tested much since I have no sm_75 capable GPU.)

rickardp commented:
> > Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling. Unfortunately the original PRs (#220 etc) cannot be viewed because of force pushes. I'll see if I can dig up a relevant diff
> 
> Windows specific stuff extracted on PR #873 and #876 See also https://github.com/wkpark/bitsandbytes/actions (several times compiled, partial squashed rebased blah blah..)
> 
> (and pytest stuff is not fixed/tested much since I have no sm_75 capable GPU.)

I don't have a Windows machine at hand right now, but there are also specific compiler flags that MSVC needs. I think this is what PR #220 did.

I spend a few minutes to rebase and create a draft PR in #947 just for reference, which should compile on all platforms. Maybe it's worth just comparing notes on those?

niclimcy commented:
I am not sure if current versions of bitsandbytes depend on pthread but attempt #220 imports pthread-win32 is a library as pthread isn’t supported on windows either.

wkpark commented:
rebased to test the current action workflows.
https://github.com/wkpark/bitsandbytes/actions/runs/7700400727/job/20984178019 (with minimal windows fix #876)

younesbelkada reviewed (approved):
Thanks a lot @wkpark 
I am very much in favor of this PR as it currently
1- tests bnb compilation on windowns AND ubuntu on each commit on main + on each PR to make sure future PRs will not break bnb compilation
2- makes bnb compilation easier through Cmake
3- Once we merge this PR https://github.com/TimDettmers/bitsandbytes/pull/876 should be closed
I will let @Titus-von-Koeller give a final decision on this - I assume the classic `make xxx` commands will still work - @wkpark once this gets merged, would you be happy to update the way we build bnb on our docker images to leverage CMake ? We do it exactly here: https://github.com/huggingface/peft/blob/189a9a666dc90f335b6a4db77f4b856b6a8bbdfd/docker/peft-gpu-bnb-source/Dockerfile#L60 and here: https://github.com/huggingface/peft/blob/189a9a666dc90f335b6a4db77f4b856b6a8bbdfd/docker/peft-gpu-bnb-latest/Dockerfile#L60

akx commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,161 @@
+            cpp_compiler: g++
+
+    steps:
+    - uses: actions/checkout@v3
```
```suggestion
    - uses: actions/checkout@v4
```

akx commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,161 @@
+    steps:
+    - uses: actions/checkout@v3
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v3
```
```suggestion
      uses: actions/setup-python@v5
```

akx commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,161 @@
+        echo "build-output-dir=${{ github.workspace }}/build" >> "$GITHUB_OUTPUT"
+
+    - name: CUDA Toolkit
+      # not work on windows
```
Should this step then be guarded with a suitable `if:`?

wkpark replied:
I mean cuda-toolkits does not work as expected. (outdated comment)

akx commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,161 @@
+    - name: Build dist
+      shell: bash -el {0}
+      run: |
+        python setup.py bdist_wheel
```
Calling `setup.py` directly is discouraged (see discussion in #949).

akx commented on setup.py:
```diff
@@ -36,6 +36,7 @@ def read(fname):
     },
     long_description=read("README.md"),
     long_description_content_type="text/markdown",
+    ext_modules=[Extension("bitsandbytes", sources=[], language="c")],
```
Is there actually a Python extension module by the name of `bitsandbytes`?  If not, I don't think this is correct?

wkpark replied:
> Is there actually a Python extension module by the name of `bitsandbytes`? If not, I don't think this is correct?

there is no actual extension module named `bitsandbytes`, but how do we distinguish a linux bin and a win bin?
with this fix, we can obtain both `bitsandbytes-0.42.0-cp310-cp310-win_amd64.whl` and `bitsandbytes-0.42.0-cp310-cp310-linux_x86_64..whl`. without this fix, we can get only one `bitsandbytes-*-py3-none-any.whl` wheel.

akx replied:
Right, so the intent is to signal to the build machinery that there is native code in here and the wheel should be tagged as such 😄 

That wasn't clear from the (lack of 😉) comment or the commit message "add ext_modules" :)

Let's at least make the intent clear here so whoever will port this to a non-setup.py build knows what's up 😄 

```suggestion
    # HACK: pretend we have a native extension module so the wheel is tagged 
    #       correctly with a platform tag (e.g. `-linux_x86_64.whl`).
    ext_modules=[Extension("bitsandbytes", sources=[], language="c")],
```

wkpark commented:
> @TimDettmers says:
> Thank you! We are planning to switch to make, and this is a great PR!
> 
> I do not have an overview of how this is different or could benefit from #220 and #229. Could you please have a look and discuss the differences and if we could benefit from some of the old code. Thank you!

PR #220 by @niclimcy and PR #229 by @acpopescu are almost identical with PR #788 @Jamezo97
(some PRs are already closed.)
so I've cleaned up some unrelated stuff, cherry picked/partial squashed and extracted into minimal subsets of several PRs.

- pthread vs std::thread -> make pthread optional. : win32 fix PR
- do not rename `pythonInterface.cpp` : win32 fix PR
- AVX related : win32 fix PR
- cmake -> extract PR
- lib detection -> extract PR
- documents and trivial fixes -> drop

See also:
- https://github.com/TimDettmers/bitsandbytes/pull/220/commits/4c1d3089cc8d1f1644f334b18402a45c1ff65a44
- https://github.com/TimDettmers/bitsandbytes/pull/229/commits/68da6d806f33497013e1211d1040840768ff75ea

wkpark commented:
Partially squashed, updated.

matthewdouglas commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,166 @@
+        -DCMAKE_CXX_COMPILER=${{ matrix.cpp_compiler }}
+        -DCMAKE_C_COMPILER=${{ matrix.c_compiler }}
+        -DCMAKE_BUILD_TYPE=${{ matrix.build_type }}
+        -DCOMPUTE_CAPABILITY="60;61;62;70;72;75;80;86;87;89;90"
```
Is it intentional to drop Maxwell here?

wkpark replied:
> Is it intentional to drop Maxwell here?

that's not intended. I just removed some computes to reduce compile time a bit.

younesbelkada commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,165 @@
+      fail-fast: false
+
+      matrix:
+        os: [ubuntu-latest]
```
```suggestion
        os: [ubuntu-latest, windows-latest]
```

younesbelkada reviewed (approved):
Thanks @wkpark for your great work ! as #876 being merged could you re-enable windows build in the workflow + rebase with main?

younesbelkada commented:
@wkpark could you also update the documentation and the main reamde with instructions on how to build bnb with windows?

younesbelkada commented:
For the docs you can start filling some content here: https://github.com/TimDettmers/bitsandbytes/blob/main/docs/source/installation.mdx

wkpark commented:
- [x] fixed mistake d97700af6 in PR #876
- [x] add windows build matrix

younesbelkada commented:
@wkpark Thanks a lot ! let me know if you need any help for https://github.com/TimDettmers/bitsandbytes/pull/908#issuecomment-1920428847 - could you just share how to build bnb from source using a windows machine, i.e, the exact commands and I'll push on your branch

wkpark commented:
> For the docs you can start filling some content here: https://github.com/TimDettmers/bitsandbytes/blob/main/docs/source/installation.mdx

this is just a note for windows user who wanted to build from source.

1. git clone
2. `cmake -B` ***build*** `-DBUILD_CUDA=ON -S .`
3. `cmake --build` ***build*** `--config Release`
4. `python -m build --wheel` (build python wheel using `build` module.) then you can get `dist/*.whl`

younesbelkada commented:
@wkpark check out https://moon-ci-docs.huggingface.co/docs/bitsandbytes/pr_908/en/installation now there is a nice table that showcases how to install bnb: 

<img width="925" alt="Screenshot 2024-02-01 at 16 16 56" src="https://github.com/TimDettmers/bitsandbytes/assets/49240599/f00b3da1-0f9c-42bd-8ba7-0ca0255ae055">

<img width="925" alt="Screenshot 2024-02-01 at 16 17 17" src="https://github.com/TimDettmers/bitsandbytes/assets/49240599/44b31842-cc47-4733-b3e9-1b449b8399aa">

younesbelkada commented on docs/source/installation.mdx:
````diff
@@ -1,3 +1,43 @@
+python -m build --wheel
+```
+
+Big thanks to [wkpark](https://github.com/wkpark), [Jamezo97](https://github.com/Jamezo97), [rickardp](https://github.com/rickardp), [akx](https://github.com/akx) for their amazing contribution to make bitsandbytes compatible with Windows.
````
```suggestion
Big thanks to [wkpark](https://github.com/wkpark), [Jamezo97](https://github.com/Jamezo97), [rickardp](https://github.com/rickardp), [akx](https://github.com/akx) for their amazing contributions to make bitsandbytes compatible with Windows.
```

younesbelkada commented:
Thanks again for all your awesome contributions ! @rickardp @wkpark @akx @Jamezo97 !

Neumann-A commented on CMakeLists.txt:
```diff
@@ -0,0 +1,132 @@
+
+
+if(BUILD_CUDA)
+    target_compile_definitions(bitsandbytes PUBLIC BUILD_CUDA)
+    target_link_libraries(bitsandbytes PUBLIC cudart cublas cusparse)
+    if(NO_CUBLASLT)
+        target_compile_definitions(bitsandbytes PUBLIC NO_CUBLASLT)
+    else()
+        target_link_libraries(bitsandbytes PUBLIC cublasLt)
+    endif()
```
Should have used targets from https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html . This now assumes that the libraries are on the library search path

akx replied:
All of this will be tweaked soon anyway, don't worry – see #949 for the follow-up.
